src/ralph/backends/data/es.py

```diff
@@ -0,0 +1,320 @@
+                instead. See `BaseOperationType`.
+
+        Returns:
+            int: The number of written documents.
+
+        Raises:
+            BackendException: If a failure occurs while writing to Elasticsearch or
+                during document decoding and `ignore_errors` is set to `False`.
+            BackendParameterException: If the `operation_type` is `APPEND` as it is not
+                supported.
+        """
+        count = 0
+        data = iter(data)
+        try:
+            first_record = next(data)
+        except StopIteration:
+            logger.debug("Data Iterator is empty; skipping write to target.")
+            return count
+        if not operation_type:
+            operation_type = self.default_operation_type
+        target = target if target else self.default_index
+        if operation_type == BaseOperationType.APPEND:
+            msg = "Append operation_type is not supported."
+            logger.error(msg)
+            raise BackendParameterException(msg)
+
+        data = chain((first_record,), data)
+        if isinstance(first_record, bytes):
+            data = self._parse_bytes_to_dict(data, ignore_errors)
+
+        logger.debug(
+            "Start writing to the %s index (chunk size: %d)", target, chunk_size
+        )
+        try:
+            for success, action in streaming_bulk(
+                client=self.client,
+                actions=self.to_documents(data, target, operation_type),
+                chunk_size=chunk_size,
+                raise_on_error=(not ignore_errors),
+            ):
+                count += success
+                logger.debug(
+                    "Wrote %d documents [action: %s ok: %d]", count, action, success
+                )
+        except (BulkIndexError, ApiError, ESConnectionError) as error:
+            raise BackendException(*error.args, f"{count} succeeded writes") from error
+        return count
+
+    @staticmethod
+    def to_documents(
+        data: Iterable[dict],
+        target: str,
+        operation_type: BaseOperationType,
+    ) -> Iterator[dict]:
+        """Converts dictionaries from `data` to ES documents and yields them."""
+        if operation_type == BaseOperationType.UPDATE:
+            for item in data:
+                yield {
+                    "_index": target,
+                    "_id": item.get("id", None),
+                    "_op_type": operation_type.value,
+                    "doc": item,
+                }
+        elif operation_type in (BaseOperationType.CREATE, BaseOperationType.INDEX):
+            for item in data:
+                yield {
+                    "_index": target,
+                    "_id": item.get("id", None),
+                    "_op_type": operation_type.value,
+                    "_source": item,
+                }
+        else:
+            # operation_type == BaseOperationType.DELETE (by exclusion)
```

Comment:
What about BaseOperationType.APPEND?

Reply:
Good point, `BaseOperationType.APPEND` is already excluded in the `write` method.
However, if `to_documents` is called in another context, indeed, this might occur.
Should we make this method private?
Or maybe move the:
```
if operation_type == BaseOperationType.APPEND:
            msg = "Append operation_type is not supported."
            logger.error(msg)
            raise BackendParameterException(msg)
```
from the `write` method in `to_documents`?

Reply:
Thanks) yes, let's make the `to_documents` private, updated it)